backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java

```diff
@@ -1051,4 +1062,77 @@ public ZoneId getTimeZone(@NonNull final I_C_Order order)
 		final OrgId orgId = OrgId.ofRepoIdOrAny(order.getAD_Org_ID());
 		return orgsRepo.getTimeZone(orgId);
 	}
+
+	@Override
+	public void validateHaddexOrder(final I_C_Order order)
+	{
+		final IOrderDAO orderDAO = Services.get(IOrderDAO.class);
+		final IProductBL productBL = Services.get(IProductBL.class);
+
+		if (!isHaddexOrder(order))
+		{
+			return;
+		}
+
+		boolean hasHaddexLine = orderDAO.retrieveOrderLines(order)
+				.stream()
+				.anyMatch(lineId -> productBL.isHaddexProduct(ProductId.ofRepoId(lineId.getM_Product_ID())));
+
+		if (!hasHaddexLine)
+		{
+			return;
+		}
+
+		validateHaddexDate(order);
+	}
+
+	@Override
+	public void validateHaddexDate(final I_C_Order order)
+	{
+		final IBPartnerDAO partnerDAO = Services.get(IBPartnerDAO.class);
+		final org.compiere.model.I_C_BPartner partner = partnerDAO.getById(order.getC_BPartner_ID());
+		final long differenceBetweenHaddexCheckDateAndPromisedDateInMonths = Math.abs(
+				ChronoUnit.MONTHS.between(
+						TimeUtil.asZonedDateTime(partner.getDateHaddexCheck()),
+						TimeUtil.asZonedDateTime(order.getDatePromised())
+				));
+
+		if (differenceBetweenHaddexCheckDateAndPromisedDateInMonths > getMaxHaddexAgeInMonths(order.getAD_Client_ID(), order.getAD_Org_ID()))
+		{
+			throw new AdempiereException(MSG_HADDEX_CHECK_ERROR).markAsUserValidationError();
+		}
+	}
+
+	@Override
+	public boolean isHaddexOrder(final I_C_Order order)
+	{
+		if (!order.isSOTrx())
+		{
+			return false;
+		}
+
+		final org.compiere.model.I_C_BPartner partner = partnerDAO.getById(order.getC_BPartner_ID());
+
+		if (!partner.isHaddexCheck())
+		{
+			return false;
+		}
+
+		if (partner.getDateHaddexCheck() == null)
```

Comment:
are we sure this is what we want? what not insisten on a date if isHaddexCheck=true?

Reply:
to be handled in a followup